src/bigint.rs

```diff
@@ -467,146 +522,303 @@ impl Unsigned for BigUint {}
+         *
+         *       = -(x0 * y0 - x0 * y1 - x1 * y0 + x1 * y1) + p0 + p2
+         *
+         *       = -((x1 - x0) * (y1 - y0)) + p0 + p2
+         *
+         * Let p1 = (x1 - x0) * (y1 - y0), and substitute back into our original formula:
+         *
+         * x * y = p0
+         *       + (p0 + p2 - p1) * b
+         *       + p2 * b^2
+         *
+         * Where the three intermediate products are:
+         *
+         * p0 = x0 * y0
+         * p1 = (x1 - x0) * (y1 - y0)
+         * p2 = x1 * y1
+         *
+         * In doing the computation, we take great care to avoid unnecessary temporary variables
+         * (since creating a BigUint requires a heap allocation): thus, we rearrange the formula a
+         * bit so we can use the same temporary variable for all the intermediate products:
+         *
+         * x * y = p2 * b^2 + p2 * b
+         *       + p0 * b + p0
+         *       - p1 * b
+         *
+         * The other trick we use is instead of doing explicit shifts, we slice acc at the
+         * appropriate offset when doing the add.
+         */
+
+        /*
+         * When x is smaller than y, it's significantly faster to pick b such that x is split in
+         * half, not y:
+         */
+        let b = x.len() / 2;
+        let (x0, x1) = x.split_at(b);
+        let (y0, y1) = y.split_at(b);
+
+        /* We reuse the same BigUint for all the intermediate multiplies: */
+
+        let len = y.len() + 1;
+        let mut p: BigUint = BigUint { data: Vec::with_capacity(len) };
+        p.data.extend(repeat(0).take(len));
+
+        // p2 = x1 * y1
+        mac3(&mut p.data[..], x1, y1);
+
+        // Not required, but the adds go faster if we drop any unneeded 0s from the end:
+        p = p.normalize();
+
+        add2(&mut acc[b..],        &p.data[..]);
+        add2(&mut acc[b * 2..],    &p.data[..]);
+
+        // Zero out p before the next multiply:
+        p.data.truncate(0);
+        p.data.extend(repeat(0).take(len));
+
+        // p0 = x0 * y0
+        mac3(&mut p.data[..], x0, y0);
+        p = p.normalize();
+
+        add2(&mut acc[..],                &p.data[..]);
+        add2(&mut acc[b..],        &p.data[..]);
+
+        // p1 = (x1 - x0) * (y1 - y0)
+        // We do this one last, since it may be negative and acc can't ever be negative:
+        let j0 = sub_sign(x1, x0);
+        let j1 = sub_sign(y1, y0);
+
+        match j0.sign * j1.sign {
+            Plus    => {
+                p.data.truncate(0);
+                p.data.extend(repeat(0).take(len));
+
+                mac3(&mut p.data[..], &j0.data.data[..], &j1.data.data[..]);
+                p = p.normalize();
+
+                sub2(&mut acc[b..], &p.data[..]);
+            },
+            Minus   => {
+                mac3(&mut acc[b..], &j0.data.data[..], &j1.data.data[..]);
+            },
+            NoSign  => (),
+        }
+    }
+}
+
+fn mul3(x: &[BigDigit], y: &[BigDigit]) -> BigUint {
```

Comment:
On Mon, Nov 30, 2015 at 03:42:11PM -0800, Josh Stone wrote:
> > +                p.data.extend(repeat(0).take(len));
> > +
> > +                mac3(&mut p.data[..], &j0.data.data[..], &j1.data.data[..]);
> > +                p = p.normalize();
> > +
> > +                sub2(&mut acc[b..], &p.data[..]);
> > +            },
> > +            Minus   => {
> > +                mac3(&mut acc[b..], &j0.data.data[..], &j1.data.data[..]);
> > +            },
> > +            NoSign  => (),
> > +        }
> > +    }
> > +}
> > +
> > +fn mul3(x: &[BigDigit], y: &[BigDigit]) -> BigUint {
> 
> Just curious, what is the 3 in `mul3`?  But anyway, this function is only used by `Mul::mul`, which itself isn't doing anything else, which raises the question why this should be separate at all.

Back when I still had an add3 and needed to distinguish, I was just using the
same convention you use in assembly for talking about ops that overwrite one of
the arguments for the result, or have another argument for the destination (e.g.
two argument add/mul that overwrites an argument for the dest, or 3 argument
that doesn't - or mac3 vs. mac4 (multiply accumulate)). It probably doesn't make
much sense anymore but it still aesthetically pleases me somehow :)

Though if we do ever (hopefully) get alloca() or some other way of moving heap
allocations to the stack, it could make sense to add a mul2() - avoid doing any
heap allocations by putting all the temporaries (including the result) on the
stack, and then return the result by copying over one of the (heap allocated)
arguments.

It does also get used in other patches I have (e.g. divide and conquer division
is able to make use of multiplying slices).